[FLINK-25756] [connectors/opensearch] Dedicated Opensearch connectors

reta opened the pull request:
Signed-off-by: Andriy Redko <andriy.redko@aiven.io>

<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

The goal of this change is to provide dedicated Opensearch connectors [1], [2], [3], [4].

## Brief change log

The implementation is largely based on the existing Elasticsearch 7 connector with a few notable changes (besides the dependencies and APIs):
 - any mentions and uses of mapping types have been removed: it is deprecated feature, scheduled for removal (the indices with mapping types cannot be created or migrated to Opensearch 1.x and beyond)
 - any mentions and uses have been removed: it is deprecated feature, scheduled for removal (only `HighLevelRestClient` is used)
 - the default distributions of Opensearch come with HTTPS turned on, using self-signed certificates: to simplify the integration a new option `allow-insecure` has been added to suppress certificates validation for development and testing purposes
 - old streaming APIs are also supported to facilitate the migration of existing applications from Elasticsearch 7/6 to Opensearch (the classes will change but the familiar model will stay)

The new connector name is `opensearch` and it follows the existing conventions:

```
CREATE TABLE users ( ... ) WITH (
  'connector' = 'opensearch', 
  'hosts' = 'https://localhost:9200',
  'index' = 'users', 
  'allow-insecure' = 'true', 
  'username' = 'admin', 
  'password' = 'admin');
```

## Verifying this change

This change added comprehensive tests and can be verified as follows (largely ported the existing unit and integration tests for Elasticsearch 7):
  - Added unit tests
  - Added integration tests for end-to-end
  - Added end-to-end tests
  - Manually verified the connector by running a node clusters

## Does this pull request potentially affect one of the following parts:

  - Dependencies: yes (the latest Opensearch 1.2.4 APIs as of this moment)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? (docs - in progress, JavaDocs)

Huge thanks @snuyanzin for help.
Retargeting https://github.com/apache/flink/pull/18541 to separate repository

[1] https://cwiki.apache.org/confluence/display/FLINK/FLIP-243%3A+Dedicated+Opensearch+connectors
[2] https://www.mail-archive.com/dev@flink.apache.org/msg58911.html
[3] https://lists.apache.org/thread/jls0vqc7jb84jp14j4jok1pqfgo2cl30
[4] https://lists.apache.org/thread/4bms24983g38q956rp8qmm4bpdo4361s

reta commented:
@MartijnVisser could you please take a look? thank you :-)

reta commented:
@MartijnVisser doing my deligance by pinging once per month :-)

reta commented:
@MartijnVisser the POMs and CI scripts have been updated accordingly

zentol commented on flink-connector-opensearch-e2e-tests/pom.xml:
```diff
@@ -0,0 +1,104 @@
+		<groupId>org.apache.flink</groupId>
+		<artifactId>flink-connector-opensearch-parent</artifactId>
+		<version>1.0.0-SNAPSHOT</version>
+		<relativePath>..</relativePath>
```
```suggestion
```

zentol commented on flink-connector-opensearch-e2e-tests/pom.xml:
```diff
@@ -0,0 +1,104 @@
+							<version>${flink.version}</version>
+							<destFileName>flink-connector-test-utils.jar</destFileName>
+							<type>jar</type>
+							<outputDirectory>${project.build.directory}/dependencies
+							</outputDirectory>
```
```suggestion
							<outputDirectory>${project.build.directory}/dependencies</outputDirectory>
```

zentol commented on flink-connector-opensearch-e2e-tests/pom.xml:
```diff
@@ -0,0 +1,104 @@
+						</goals>
+					</execution>
+				</executions>
+				<configuration>
```
can we move this configuration into the execution?

reta replied:
> can we move this configuration into the execution?

Sure, may need need same adjustment here https://github.com/apache/flink-connector-elasticsearch/blob/main/flink-connector-elasticsearch-e2e-tests/flink-connector-elasticsearch7-e2e-tests/pom.xml#L109 and https://github.com/apache/flink-connector-elasticsearch/blob/main/flink-connector-elasticsearch-e2e-tests/flink-connector-elasticsearch6-e2e-tests/pom.xml#L108

zentol commented on flink-connector-opensearch/pom.xml:
```diff
@@ -0,0 +1,185 @@
+
+	<!-- Allow users to pass custom connector versions -->
+	<properties>
+		<opensearch.version>1.3.0</opensearch.version>
```
Can you give us some background of the opensearch versioning?
I see 2.3.0 artifacts on maven central.

reta replied:
> Can you give us some background of the opensearch versioning? I see 2.3.0 artifacts on maven central.

Sure, so we have 2 maintainable release trains: 1.x and 2.x. The major difference between two (leaving all features aside) is that 1.x has JDK-8 baseline while 2.x has JDK-11 baseline. Since Flink still has JDK-8 baseline, we are using 1.x client BUT it will work with 2.x servers.

zentol commented on flink-connector-opensearch/pom.xml:
```diff
@@ -0,0 +1,185 @@
+
+		<!-- Core -->
+
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-connector-base</artifactId>
+			<version>${flink.version}</version>
+		</dependency>
```
best to set this to provided to avoid compatibility issues between connector-base and Flink.

reta replied:
> best to set this to provided to avoid compatibility issues between connector-base and Flink.

Needs adjustment here https://github.com/apache/flink-connector-elasticsearch/blob/main/flink-connector-elasticsearch-base/pom.xml#L47

zentol commented on flink-connector-opensearch/pom.xml:
```diff
@@ -0,0 +1,185 @@
+		<groupId>org.apache.flink</groupId>
+		<artifactId>flink-connector-opensearch-parent</artifactId>
+		<version>1.0.0-SNAPSHOT</version>
+		<relativePath>..</relativePath>
```
```suggestion
```

zentol commented on flink-connector-opensearch/pom.xml:
```diff
@@ -0,0 +1,185 @@
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-surefire-plugin</artifactId>
+				<configuration>
+					<!-- Enforce single fork execution because of spawning
+					 Opensearch cluster multiple times -->
+					<forkCount>1</forkCount>
```
IIRC we got rid of this in the elasticsearch connector, so we shouldn't need this.

zentol commented on flink-sql-connector-opensearch/pom.xml:
```diff
@@ -0,0 +1,159 @@
+		<groupId>org.apache.flink</groupId>
+		<artifactId>flink-connector-opensearch-parent</artifactId>
+		<version>1.0.0-SNAPSHOT</version>
+		<relativePath>..</relativePath>
```
```suggestion
```

zentol commented on flink-sql-connector-opensearch/pom.xml:
```diff
@@ -0,0 +1,159 @@
+								<filter>
+									<artifact>*:*</artifact>
+									<excludes>
+										<!-- exclude Java 9 specific classes as otherwise the shade-plugin crashes -->
+										<exclude>META-INF/versions/**</exclude>
```
This seems odd; Flink itself is bundling java version-specific files.

Can you show me the error message you got?

reta replied:
This one is taken from https://github.com/apache/flink-connector-elasticsearch/blob/main/flink-sql-connector-elasticsearch7/pom.xml#L117 since OpenSearch is a fork of Elasticseach 7.10.2

zentol replied:
What happens without it? :see_no_evil:

reta replied:
Could be quite outdated, removed (for now)

reta replied:
> What happens without it? see_no_evil

@zentol answer to your (and mine) curiosity:

```
 Error:  Failed to execute goal org.apache.maven.plugins:maven-shade-plugin:3.1.1:shade (shade-flink) on project flink-sql-connector-opensearch: Error creating shaded jar: Problem shading JAR /home/runner/.m2/repository/org/opensearch/opensearch/1.3.0/opensearch-1.3.0.jar entry META-INF/versions/11/org/opensearch/monitor/jvm/JvmPid.class: java.lang.IllegalArgumentException -> [Help 1]
``

zentol replied:
Alright then. :sweat_smile:

zentol commented on flink-sql-connector-opensearch/pom.xml:
```diff
@@ -0,0 +1,159 @@
+								</filter>
+							</filters>
+							<relocations>
+								<!-- Force relocation of all Opensearch dependencies. -->
```
```suggestion
```

zentol commented on flink-sql-connector-opensearch/src/main/resources/META-INF/NOTICE:
```diff
@@ -0,0 +1,50 @@
+flink-sql-connector-opensearch
+Copyright 2014-2021 The Apache Software Foundation
```
```suggestion
Copyright 2014-2022 The Apache Software Foundation
```

zentol reviewed (changes requested):
Made a pass over infrastructure/build/legal files.

reta commented:
> Can you split the commit such that 1 adds the workflows and another the rest? We can then merge the workflow commit first and actually get CI in this PR.

Sure, https://github.com/apache/flink-connector-opensearch/pull/2, thanks @zentol

zentol commented on flink-connector-opensearch-e2e-tests/pom.xml:
```diff
@@ -0,0 +1,190 @@
+
+	<profiles>
+		<profile>
+			<id>run-end-to-end-tests</id>
```
There is no longer a real benefit in not treating the e2e tests as normal IT Cases.
(notably because we always run them on CI and running them locally works just fine as well)

reta replied:
@zentol this it not correct, the e2e tests need Flink distribution to be downloaded and provided, this preparation is done within CI workflows, otherwise the tests won't even start.

zentol replied:
That's a good point.

We should fix that though (in a follow-up; across connectors); the tests should be able to download Flink themselves. These magical prerequisite steps just make it difficult to test things locally.

reta replied:
:+1: second that. It is, took me a while to understand how I could run it locally

zentol commented on flink-connector-opensearch-e2e-tests/pom.xml:
```diff
@@ -0,0 +1,190 @@
+									<includes>
+										<include>**/*.*</include>
+									</includes>
+									<!-- E2E tests must not access flink-dist concurrently. -->
+									<forkCount>1</forkCount>
```
This shouldn't be required anymore.

zentol commented on flink-connector-opensearch-e2e-tests/pom.xml:
```diff
@@ -0,0 +1,190 @@
+						<configuration>
+							<finalName>opensearch-end-to-end-test</finalName>
+							<outputDirectory>dependencies</outputDirectory>
+							<artifactSet>
+								<excludes combine.children="append">
+									<exclude>com.google.code.findbugs:jsr305</exclude>
+									<exclude>org.slf4j:slf4j-api</exclude>
+								</excludes>
+							</artifactSet>
+							<filters combine.children="append">
+								<filter>
+									<artifact>*:*</artifact>
+									<excludes>
+										<exclude>META-INF/*.SF</exclude>
+										<exclude>META-INF/*.DSA</exclude>
+										<exclude>META-INF/*.RSA</exclude>
+									</excludes>
+								</filter>
+							</filters>
```
If you'd set execution id to `shade-flink` this wouldn't be necessary.

reta replied:
It seems like even `shade-flink` is not needed, removed all that

zentol commented on pom.xml:
```diff
@@ -327,6 +327,12 @@ under the License.
 			<!-- For dependency convergence -->
+			<dependency>
+				<groupId>net.java.dev.jna</groupId>
+				<artifactId>jna</artifactId>
```
Where this come from?

reta replied:
Dependency convergence, the OpenSearch brings older version

zentol replied:
So could we exclude it in general? Or do we need it for some tests to work?

What is the bootstrap licensing issue about?

zentol commented on flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSinkBuilder.java:
```diff
@@ -0,0 +1,449 @@
+        return self();
+    }
+
+    protected BulkProcessorBuilderFactory getBulkProcessorBuilderFactory() {
```
There's no reason for this to be protected since BulkProcessorBuilderFactory is an internal interface. 
Maybe that entire class shouldn't exist in it's current form, as it only existed to supported different elasticsearch versions (unless we intend to do the same in the near future for opensearch).

reta replied:
> There's no reason for this to be protected since BulkProcessorBuilderFactory is an internal interface

:+1: 

> Maybe that entire class shouldn't exist in it's current form, as it only existed to supported different elasticsearch versions (unless we intend to do the same in the near future for opensearch).

Yes, it could be removed since the only place `BulkProcessor` is need is inside the `OpensearchWriter`

zentol commented on pom.xml:
```diff
@@ -0,0 +1,463 @@
+
+	<modules>
+		<module>flink-connector-opensearch</module>
+		<module>flink-connector-opensearch-e2e-tests</module>
```
missing the sql connector module

reta replied:
It is not missing, driven by profile [1] which is active by default. Reason: it comes from Flink itself [2]

[1] https://github.com/apache/flink-connector-opensearch/pull/1/files/41cc7bb681aede94c543c28cc5629802874aac70#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8R385
[2] https://github.com/apache/flink/blob/master/flink-connectors/pom.xml#L83

reta replied:
Sure, thanks for clarification, removing the profile

zentol commented on flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSink.java:
```diff
@@ -0,0 +1,91 @@
+ * @see OpensearchSinkBuilder on how to construct a OpensearchSink
+ */
+@PublicEvolving
+public class OpensearchSink<IN> implements Sink<IN> {
```
Curious why this class lost the `getDeliveryGuarantee()`

reta replied:
If you mean the equivalent from `ElasticseachSink` :
```
    @VisibleForTesting
    DeliveryGuarantee getDeliveryGuarantee() {
        return deliveryGuarantee;
    }
```

The default for `deliveryGuarantee` was changed recently and the method was added to tests specifically for that.

zentol commented on flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/NetworkClientConfig.java:
```diff
@@ -0,0 +1,86 @@
+        return connectionPathPrefix;
+    }
+
+    @Nullable
+    public Boolean isAllowInsecure() {
+        return allowInsecure;
+    }
```
Consider improving on the existing code and returning an Optional instead.

reta replied:
:+1:

zentol commented on flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/table/OpensearchConfiguration.java:
```diff
@@ -0,0 +1,168 @@
+        return config.getOptional(BULK_FLUSH_BACKOFF_DELAY_OPTION).map(Duration::toMillis);
+    }
+
+    public boolean isDisableFlushOnCheckpoint() {
```
unused. I also cant find the corresponding option in the ES connector; where does it come from?

reta replied:
It was there I believe right around the time we started to work on it [1] last year

[1] https://github.com/apache/flink/commit/6609570e49901d83911ec68ced6465f49658c791

zentol commented on flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/table/OpensearchConnectorOptions.java:
```diff
@@ -0,0 +1,161 @@
+                    .noDefaultValue()
+                    .withDescription("Delay between each backoff attempt.");
+
+    public static final ConfigOption<Boolean> FLUSH_ON_CHECKPOINT_OPTION =
```
effectively unused

zentol commented on flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/table/OpensearchDynamicSink.java:
```diff
@@ -0,0 +1,201 @@
+    final EncodingFormat<SerializationSchema<RowData>> format;
+    final DataType physicalRowDataType;
+    final List<LogicalTypeWithIndex> primaryKeyLogicalTypesWithIndex;
+    final OpensearchConfiguration config;
```
The lack of the localTimeZoneId field implies that FLINK-24571 isn't contained in this PR. I'm wondering why that is.

zentol commented on flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/table/OpensearchDynamicSinkFactory.java:
```diff
@@ -0,0 +1,222 @@
+    private final OpensearchSinkBuilderSupplier<RowData> sinkBuilderSupplier;
+
+    public OpensearchDynamicSinkFactory() {
+        this.sinkBuilderSupplier = OpensearchSinkBuilder<RowData>::new;
```
```suggestion
        this.sinkBuilderSupplier = OpensearchSinkBuilder::new;
```
not required

zentol replied:
This field doesn't seem necessary in the first place; we could handle that within the sink.

reta replied:
> This field doesn't seem necessary in the first place; we could handle that within the sink.

Not so sure, the tradeoff is to create supplier instance only once, vs creating it all the time in place

zentol replied:
If OpensearchDynamicSink#getSinkRuntimeProvider directly creates the opensearch sink builder without going through the supplier then it should be functionally identical.
(You don't need a supplier at all.)

zentol commented on flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/table/OpensearchDynamicSinkFactory.java:
```diff
@@ -0,0 +1,222 @@
+                sinkBuilderSupplier);
+    }
+
+    OpensearchConfiguration getConfiguration(Context context) {
```
can be private, probably also applies to other methods in here.

zentol commented on flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/table/OpensearchDynamicSinkFactory.java:
```diff
@@ -0,0 +1,222 @@
+        }
+    }
+
+    EncodingFormat<SerializationSchema<RowData>> getValidatedEncodingFormat(
```
Where does this come from?

reta replied:
Was there before FLINK-25391, changed implementation

zentol commented on flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/table/OpensearchDynamicSinkFactory.java:
```diff
@@ -0,0 +1,222 @@
+    }
+
+    @Override
+    public Set<ConfigOption<?>> optionalOptions() {
```
Why is this missing `forwardOptions()` (FLINK-25391)?

zentol reviewed (commented):
Do you, by chance, have a branch that shows the _actual_ changes made on the ES connector code?
Otherwise I'll just have to continue manually comparing them one-by-one and coming back to you on any differences found.

zentol commented on flink-connector-opensearch/src/test/java/org/apache/flink/streaming/connectors/opensearch/OpensearchSinkITCase.java:
```diff
@@ -0,0 +1,181 @@
+
+    private OpensearchSink<Tuple2<Integer, String>> createOpensearchSinkForNode(
+            int bulkFlushMaxActions,
+            String clusterName,
```
clusterName is unused.

zentol commented on flink-connector-opensearch/src/main/java/org/apache/flink/streaming/connectors/opensearch/OpensearchSink.java:
```diff
@@ -0,0 +1,807 @@
+        }
+    }
+
+    long getNumPendingRequests() {
```
Indeed that entire test class appears to be missing.

reta replied:
> add integration tests instead (to keep the API contained)

This is done now

zentol commented on flink-connector-opensearch-e2e-tests/src/test/java/org/apache/flink/streaming/tests/OpensearchSinkE2ECase.java:
```diff
@@ -0,0 +1,117 @@
+    @TestEnv
+    protected FlinkContainerTestEnvironment flink = new FlinkContainerTestEnvironment(1, 6);
+
+    public OpensearchSinkE2ECase() throws Exception {}
+
```
```suggestion
```

reta replied:
This is needed, the reason for that is that initialization for fields in this class could throw `Exception` (which is not allowed by default), the default constructor with `throws Exception` spec hints the compiler

zentol commented on flink-connector-opensearch-e2e-tests/src/test/java/org/apache/flink/streaming/tests/OpensearchSinkE2ECase.java:
```diff
@@ -0,0 +1,117 @@
+    CheckpointingMode[] semantics = new CheckpointingMode[] {CheckpointingMode.EXACTLY_ONCE};
+
+    @TestEnv
+    protected FlinkContainerTestEnvironment flink = new FlinkContainerTestEnvironment(1, 6);
```
```suggestion
    FlinkContainerTestEnvironment flink = new FlinkContainerTestEnvironment(1, 6);
```

reta commented on flink-connector-opensearch/src/main/java/org/apache/flink/streaming/connectors/opensearch/OpensearchSink.java:
```diff
@@ -0,0 +1,807 @@
+     * @return The created client.
+     * @throws IOException IOException
+     */
+    private RestHighLevelClient createClient(Map<String, String> clientConfig) throws IOException {
```
:+1: Removed

reta commented on flink-connector-opensearch/src/main/java/org/apache/flink/streaming/connectors/opensearch/util/IgnoringFailureHandler.java:
```diff
@@ -0,0 +1,38 @@
+
+/** Ignores all kinds of failures and drops the affected {@link ActionRequest}. */
+@Internal
+public class IgnoringFailureHandler implements ActionRequestFailureHandler {
```
I believe these 2 handlers (IgnoringFailureHandler and RetryRejectedExecutionFailureHandler) are provided as an alternatives for configuring OpensearchSink::setFailureHandler

zentol replied:
Yes and no. A user _could_ set the failure handler, but this implementation here is marked as `@Internal`; presumably because they were targeting the Table API connector.
See `ElasticsearchConfiguration#getFailureHandler`

_If_ we want to be be a drop-in replacement for the ES table api connector, then we'd need to pull in all of FLINK-26638, which also includes going back to the old ES sink for the Table API.

It's a bit of an unsolved question on the ES connector side as to what will happen with these failure handlers :/

reta replied:
Sure, thanks, I will drop them

reta commented on flink-connector-opensearch/src/main/java/org/apache/flink/streaming/connectors/opensearch/util/RetryRejectedExecutionFailureHandler.java:
```diff
@@ -0,0 +1,60 @@
+ */
+@Deprecated
+@PublicEvolving
+public class RetryRejectedExecutionFailureHandler implements ActionRequestFailureHandler {
```
Same as above - for users to use

reta commented on flink-connector-opensearch/src/test/java/org/apache/flink/connector/opensearch/table/IndexGeneratorTest.java:
```diff
@@ -0,0 +1,264 @@
+            IndexGeneratorFactory.createIndexGenerator(
+                    "my-index-{local_date|yyyy/MM/dd HH:mm}", fieldNames, dataTypes);
+        } catch (UnsupportedTemporalTypeException e) {
+            Assertions.assertEquals(expectedExceptionMsg, e.getMessage());
```
That's a big one (as a change), will try to tackle it at some point this week

reta commented:
> The architecture tests are missing. Have a look at the ES connector.

Added

zentol commented on flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSink.java:
```diff
@@ -88,4 +88,8 @@ public SinkWriter<IN> createWriter(InitContext context) throws IOException {
                 context.getMailboxExecutor());
     }
+
+    DeliveryGuarantee getDeliveryGuarantee() {
```
add `@VisibleForTesting`

zentol replied:
The architecture tests verify that these methods are only called in tests ;) It allows us to leak some implementation details into tests without impacting production code. Can make your life a lot easier.

reta replied:
I believe in APIs :) I think there are number of ways to approach testing of implementation details, but that's certainly very opinionated subjects, thanks for sharing your point of view.

zentol commented on flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/RequestIndexer.java:
```diff
@@ -0,0 +1,56 @@
+ * Users add multiple delete, index or update requests to a {@link RequestIndexer} to prepare them
+ * for sending to an Opensearch cluster.
+ */
+@Internal
```
This shouldn't be internal since it's returned by OpensearchEmitter which is PublicEvolving.
(as reported by the architecture tests ;) )

zentol commented on flink-connector-opensearch/src/test/java/org/apache/flink/architecture/TestCodeArchitectureTest.java:
```diff
@@ -0,0 +1,57 @@
+        },
+        importOptions = {
+            ImportOption.OnlyIncludeTests.class,
+            TestCodeArchitectureTest.IncludeES7ImportOption.class,
```
This isn't required; it's only needed when you have a `base` module like the ES connector does.

reta replied:
Ah ... definitely the copy/paste leftovers

zentol commented on pom.xml:
```diff
@@ -0,0 +1,447 @@
+		<flink.shaded.version>15.0</flink.shaded.version>
+		
+		<jackson-bom.version>2.13.4.20221013</jackson-bom.version>
+		<junit4.version>4.13.2</junit4.version>
```
JUnit4 should no longer be required.

reta commented on flink-connector-opensearch/src/main/java/org/apache/flink/streaming/connectors/opensearch/OpensearchSink.java:
```diff
@@ -0,0 +1,807 @@
+ * @param <T> Type of the elements handled by this sink
+ */
+@Internal
+public class OpensearchSink<T> extends RichSinkFunction<T> implements CheckpointedFunction {
```
Similarly to `org.apache.flink.streaming.connectors.elasticsearch*.ElasticsearchSink`? Probably makes sense

reta commented:
@zentol thnaks a lot for the review, I think I addressed all your comments, please let me know if I missed some, thanks!

> @reta There are some unresolved comments, like [#1 (comment)](https://github.com/apache/flink-connector-opensearch/pull/1#discussion_r1029128244), [#1 (comment)](https://github.com/apache/flink-connector-opensearch/pull/1#discussion_r1029342543), [#1 (comment)](https://github.com/apache/flink-connector-opensearch/pull/1#discussion_r1029129392).

Thanks @zentol , addressed or/and answered

zentol commented on flink-connector-opensearch/src/test/java/org/apache/flink/streaming/connectors/opensearch/OpensearchSinkTest.java:
```diff
@@ -0,0 +1,562 @@
+        testHarness.processElement(new StreamRecord<>("msg-1"));
+
+        // Await for flush to be complete
+        awaitForFlushToFinish(1);
```
You can avoid this kind of busy waiting quite easily by using `OneShotLatches`.

```
    private OneShotLatch addResponse(Consumer<HttpResponse> consumer) {
        OneShotLatch oneShotLatch = new OneShotLatch();
        responses.add(
                response -> {
                    consumer.accept(response);
                    oneShotLatch.trigger();
                });
        return oneShotLatch;
    }
....
      OneShotLatch firstResponse = responses.add(
                createResponse(
                        new BulkItemResponse(
                                1,
                                OpType.INDEX,
                                new IndexResponse(
                                        new ShardId("test", "-", 0), "_doc", "1", 0, 0, 1, true))));
...
    firstResponse.await();
```

That being said, this whole "wait for server response to be consumed" approach is a bit sketchy?
We're only waiting for the server to send the response, _not_ for the connector to process it.

reta replied:
I would argue there is definitely no need for additional latch: the queue is the "latch" by itself - we know when it is drained. We could also use https://github.com/awaitility/awaitility to make the waiting logic clean.

> That being said, this whole "wait for server response to be consumed" approach is a bit sketchy?

We are dealing with async system so I think this is kind of expectable: we are waiting for some events to happen at some point (be it latch, queue or other means). The sink does not provide any means to hook inside and, for example, we could observe the "process it" part using `getNumPendingRequests` only.

reta replied:
> The point was to not wait by looping over a condition to save CI resources. 

Sure, replaced with `Condition`, should be simpler than per-response monitors.

> This is especially a problem when the condition is never fulfilled and we just churn through CPU cycles for no good reason.

All tests are timeboxed to 5 seconds

zentol replied:
> All tests are timeboxed to 5 seconds

a) In our experience such small timeouts actually cause tests to be unstable.
b) code has a tendency to travel. Arguing that "this code is fine because of that code _over there_" basically means "this code is unsafe to copy", which isn't a good state to be in.

reta replied:
The tests seem to be very stable, I was running them continuously for 2h and have not seen any flakyness.

zentol replied:
It's more of a CI problem. Time jumping ahead and things like that. :weary:

reta replied:
I think we could always update the test cases, if the instability comes in, the 5 seconds buffer is 3x buffer for what test case should take.

zentol commented on flink-connector-opensearch/src/test/java/org/apache/flink/streaming/connectors/opensearch/OpensearchSinkTest.java:
```diff
@@ -0,0 +1,562 @@
+        testHarness.processElement(new StreamRecord<>("msg"));
+
+        // current number of pending request should be 1 due to the re-add
+        assertThat(sink.getNumPendingRequests()).isEqualTo(1);
```
This assertion sometimes fails locally when the test is run in a loop.
```
expected: 1L
 but was: 0L
```

reta replied:
Aha, I think I know why, will fix that

zentol commented on flink-connector-opensearch/src/test/java/org/apache/flink/streaming/connectors/opensearch/OpensearchSinkTest.java:
```diff
@@ -0,0 +1,562 @@
+        assertThatThrownBy(snapshotThread::sync)
+                .getCause()
+                .getCause()
+                .hasMessageContaining("Unable to parse response body");
```
This sounds like an error in the test setup.

reta replied:
Not really, so the original test case is designed to verify the condition "complete bulk response fails". The Bulk API (in Elasticsearch or Opensearch) has does not support that, the failure are only part of individual bulk item [1]. The way I was able to replicate this kind of scenario is by sending invalid payload as part of bulk response but I could not control what kind of exception should be raised (hence we have `Unable to parse response body`), I could add comment about it to clarify it is by design.

[1] https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html#docs-bulk-api-example

zentol replied:
> The Bulk API (in Elasticsearch or Opensearch) does not support that

Couldn't it fail as a whole because of connection issues etc.?

In any case this should be fine then.

zentol commented on flink-connector-opensearch/src/test/java/org/apache/flink/streaming/connectors/opensearch/OpensearchSinkTest.java:
```diff
@@ -0,0 +1,562 @@
+        }
+    }
+
+    private Consumer<HttpResponse> createResponse(BulkItemResponse item) {
```
can be static

reta commented on flink-connector-opensearch/src/test/java/org/apache/flink/streaming/connectors/opensearch/OpensearchSinkTest.java:
```diff
@@ -0,0 +1,586 @@
+        };
+    }
+
+    private static void awaitForCondition(Supplier<Boolean> condition) throws InterruptedException {
```
There is only 1 place when we need that, due to the fact that `failureRequestIndexer` is not immediately re-adding the requests but does it on the next batch interval only

zentol commented on flink-connector-opensearch/src/test/java/org/apache/flink/streaming/connectors/opensearch/OpensearchSinkTest.java:
```diff
@@ -79,7 +82,15 @@ public void setUp() throws IOException {
+                            lock.lock();
+                            try {
+                                responses.poll().accept(resp);
+                                flushed.signalAll();
```
Or you could just use a OneShotLatch which makes all of this logic here a one-liner _and_ gives you more control to wait for a specific message to be consumed :shrug: 
(and you wouldn't need `awaitForCondition`)

reta replied:
`awaitForCondition` is not checking the flushing part, it is checking the `numPendingRequest`, which will be eventually updated

reta commented:
@zentol thanks on more time for the review, I think none of your comments left unattended, thanks

wbeckler commented:
It would be amazing if this could be based on the opensearch-java client rather than RHLC, since that is where all of the improvements are going. Is that how this is set up?

reta commented:
> It would be amazing if this could be based on the opensearch-java client rather than RHLC, since that is where all of the improvements are going. Is that how this is set up?

Not for now, see please https://github.com/opensearch-project/opensearch-java/issues/181 (this is what the connector uses).

zentol commented on pom.xml:
```diff
@@ -0,0 +1,440 @@
+				<artifactId>flink-test-utils</artifactId>
+				<version>${flink.version}</version>
+				<scope>test</scope>
+				<exclusions>
+					<exclusion>
+						<groupId>log4j</groupId>
+						<artifactId>log4j</artifactId>
+					</exclusion>
+					<exclusion>
+						<groupId>org.slf4j</groupId>
+						<artifactId>slf4j-log4j12</artifactId>
+					</exclusion>
+				</exclusions>
```
```suggestion
```
Since 1.16.0 this should no longer be required.

zentol reviewed (changes requested):
Just noticed that we're lacking documentation; like https://github.com/apache/flink-connector-elasticsearch/blob/main/docs/content/docs/connectors/datastream/elasticsearch.md.

reta commented:
> Just noticed that we're lacking documentation; like https://github.com/apache/flink-connector-elasticsearch/blob/main/docs/content/docs/connectors/datastream/elasticsearch.md.

Thanks @zentol, I have it ready and thought to open separate pull request (to split MDs from code), makes sense to you?

zentol commented:
ok

ypark2103 commented:
hi @zentol @reta 
I see that flink-opensearch-connector was merged to master recently but I don't see it was registered in maven repository.
When will this be added to maven repository so I can use this from pom.xml? I also don't see opensearch connector in official flink documentation. Can this library be used in Java 8 also?

reta commented:
> hi @zentol @reta I see that [flink-opensearch-connector](https://issues.apache.org/jira/browse/FLINK-opensearch-connector) was merged to master recently but I don't see it was registered in maven repository. When will this be added to maven repository so I can use this from pom.xml? 

Hey @ypark2103 , the vote was ongoing and just approved [1], expect artifacts to be available in the coming days.

[1] https://lists.apache.org/list?dev@flink.apache.org:lte=1M:Opensearch

> I also don't see opensearch connector in official flink documentation. 

The Opensearch documentations available here [2] but I am not sure how it is going to be integrated into Flink documentation, @zentol may be you could spot some light?

[2] https://github.com/apache/flink-connector-opensearch/tree/main/docs

> Can this library be used in Java 8 also?

Yes

ypark2103 commented:
@reta Is there a way we can install this library with pom.xml?
I don't see opensearch in maven depository yet like elasticsearch connector.
https://mvnrepository.com/artifact/org.apache.flink/flink-connector-elasticsearch7

reta commented:
> The artifacts you mentioned that will be available in the coming days is referring to registering this library to maven repo?

@ypark2103 Correct

MartijnVisser commented:
> The Opensearch documentations available here [2] but I am not sure how it is going to be integrated into Flink documentation, @zentol may be you could spot some light?

They will be integrated via https://github.com/apache/flink/pull/21518 when the final steps for the release process have been completed :)

ypark2103 commented:
@reta @MartijnVisser Is this Flink-Connector-Opensearch only available for flink 1.16 version? What about flink 1.11 or 1.12?

reta commented:
@ypark2103 that is correct, the Flink Opensearch Connector was following the Flink's externalization model for connectors, the necessary scaffolding is only available in Flink 1.16 and above.

ypark2103 commented:
@reta I heard java 8 was deprecated starting Flink 1.15. Does it mean I need to upgrade to java 11 to use this connector? Or can I still use java 8 for Flink 1.16?

reta commented:
> @reta I heard java 8 was deprecated starting Flink 1.15. Does it mean I need to upgrade to java 11 to use this connector? Or can I still use java 8 for Flink 1.16?

@ypark2103 the connector itself has baseline of JDK-8 so you don't need JDK-11 to use it as-is (using 1.3.x OpenSearch client), but if you plan to upgrade to JDK-11, it is even better.
